# Advanced - Pandas: Index Management
___
- EG: `df.sort_index()`
- So far, we have learned a lot about how we can manipulate Data Frames either to Filter it and then also do things like Pivoting it
- When we do Operations like this, it kind of jacks up our Index... and so we need to know, certian Methods in order to fix back our Index

In [1]:
# Imports
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt

# Loading Data
dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()

# Data Cleanup
df['job_posted_date'] = pd.to_datetime(df['job_posted_date'])

In [ ]:
# Inspecting a Sample of the Data
df.sample(5)

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
716667,Data Analyst,Data Management Analyst,Thailand,via LinkedIn,Full-time,False,Thailand,2023-09-22 02:31:20,False,False,Thailand,NaN,NaN,NaN,Gallothai,"['spreadsheet', 'excel']","{'analyst_tools': ['spreadsheet', 'excel']}"
710833,Senior Data Analyst,Senior Data Analyst - Now Hiring,"Clearwater, FL",via Snagajob,Full-time,False,"Florida, United States",2023-06-17 02:18:36,False,True,United States,NaN,NaN,NaN,Jobot,"['sql', 'tableau', 'power bi']","{'analyst_tools': ['tableau', 'power bi'], 'pr..."
491457,Software Engineer,Senior Software Engineer,"Knysna, South Africa",via LinkedIn,Full-time,False,South Africa,2023-09-13 11:19:28,False,False,South Africa,NaN,NaN,NaN,Carlysle Human Capital,"['java', 'javascript', 'typescript', 'css', 'm...","{'async': ['jira'], 'databases': ['mysql'], 'l..."
220928,Data Engineer,Big Data Engineer,Anywhere,via LinkedIn,Full-time,True,Canada,2023-07-21 22:14:46,False,False,Canada,NaN,NaN,NaN,Digitive,"['nosql', 'mongodb', 'mongodb', 'aws', 'hadoop...","{'cloud': ['aws'], 'databases': ['mongodb'], '..."
136204,Data Scientist,Entry Level Data Scientist (Remote),"Los Angeles, CA",via Snagajob,Full-time,False,"California, United States",2023-08-07 08:20:17,False,False,United States,hour,NaN,43.98,SynergisticIT,"['python', 'scala', 'django']","{'programming': ['python', 'scala'], 'webframe..."


- Inspecting the Data Frame, using the `.sample()` Method we can see the Indexes on the Left, it doesn't have a Name and everything else is off to the right
    - But this Index is obviously Numbers so let's inspect it more
    - We can do this by Invoking the `index` Property, which Accesses the Index Atribute

In [3]:
# Inspecting the Index using the index property
df.index

RangeIndex(start=0, stop=785741, step=1)

- This tells us it's a `RangeIndex` that goes from 0 to the last Number which is 785,741 and it's stepping/incrementing by 1
    - Which make sense and we can Confirm this by using the `head()` Method and seeing it Increment by 1 and seeing that it's in Numerical Order

In [4]:
# Validating the Increment and seeing it's in Numerical Order
df.head(5)

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
1,Data Analyst,Data Analyst,"Guadalajara, Jalisco, Mexico",via BeBee México,Full-time,False,Mexico,2023-01-14 13:18:07,False,False,Mexico,NaN,NaN,NaN,Hewlett Packard Enterprise,"['r', 'python', 'sql', 'nosql', 'power bi', 't...","{'analyst_tools': ['power bi', 'tableau'], 'pr..."
2,Data Engineer,"Data Engineer/Scientist/Analyst, Mid or Senior...","Berlin, Germany",via LinkedIn,Full-time,False,Germany,2023-10-10 13:14:55,False,False,Germany,NaN,NaN,NaN,ALPHA Augmented Services,"['python', 'sql', 'c#', 'azure', 'airflow', 'd...","{'analyst_tools': ['dax'], 'cloud': ['azure'],..."
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
4,Data Engineer,Data Engineer- Sr Jobs,"Washington, DC",via Clearance Jobs,Full-time,False,Sudan,2023-08-07 14:29:36,False,False,Sudan,NaN,NaN,NaN,Kristina Daniel,"['bash', 'python', 'oracle', 'aws', 'ansible',...","{'cloud': ['oracle', 'aws'], 'other': ['ansibl..."


- Another Attribute we can look at is `.name`
- This is used to see what is the name of the Index
    - Running this shows that there's nothing in there, it's Blank right now

In [5]:
# Looking at the Name of the Index
df.index.name

- Now Attributes are basically just a Variable Assigned to a particular Object
    - In this Case, the Data Frame
- So Name, is technically just like a Variable holder
    - If we wanted to, in this case, we could Assign a Name to this Attribute of "job_index"
    - So let's do just that!

In [6]:
# Assigning a Name to the Index
df.index.name = 'job_index'

In [7]:
# Printing out the Data Frame below to Inspect and see that New Name
df.head()

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
job_index,,,,,,,,,,,,,,,,,
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
1,Data Analyst,Data Analyst,"Guadalajara, Jalisco, Mexico",via BeBee México,Full-time,False,Mexico,2023-01-14 13:18:07,False,False,Mexico,NaN,NaN,NaN,Hewlett Packard Enterprise,"['r', 'python', 'sql', 'nosql', 'power bi', 't...","{'analyst_tools': ['power bi', 'tableau'], 'pr..."
2,Data Engineer,"Data Engineer/Scientist/Analyst, Mid or Senior...","Berlin, Germany",via LinkedIn,Full-time,False,Germany,2023-10-10 13:14:55,False,False,Germany,NaN,NaN,NaN,ALPHA Augmented Services,"['python', 'sql', 'c#', 'azure', 'airflow', 'd...","{'analyst_tools': ['dax'], 'cloud': ['azure'],..."
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
4,Data Engineer,Data Engineer- Sr Jobs,"Washington, DC",via Clearance Jobs,Full-time,False,Sudan,2023-08-07 14:29:36,False,False,Sudan,NaN,NaN,NaN,Kristina Daniel,"['bash', 'python', 'oracle', 'aws', 'ansible',...","{'cloud': ['oracle', 'aws'], 'other': ['ansibl..."


- Inspecting the above, we can see that we now have basically a new line inserted in here with `job_index` on here
- And we can go further now and actually Investigate the Items which is another Attribute we can look at which is the Data Type
    - This is found by using the `dtype` Attribute

In [8]:
# Looking at the dtype of the Index
df.index.dtype

dtype('int64')

- And we got Returned back that this is an Int64 or Integer
- This is because these are Integers in here for this Index

### Three Methods to Cover
___
- Now that we have that Refresher out of the way, we are going to be Focusing on diving deeper into these 3 Methods
1. `df.reset_index()`
2. `df.set_index()`
3. `df.sort_index()`
- We are going to go through explaining it in our Data Frame
- So let's jump into it and look at the first one `df.reset_index()`

### `df.reset_index()`
___
- Let's say we want to Create a New Data Frame that it's Filtered for Jobs, particular to the United States
    - We Create a New Variable with the appropriate Name, and we Set it equal to our Data Frame, Filtering it for the Job Country of USA

In [ ]:
# Creating a New Variable and Setting it to the Data Frame, Filtering it for postings in the Unites States 
df_usa = df[df['job_country'] == 'United States']
df_usa

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
job_index,,,,,,,,,,,,,,,,,
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
5,Data Engineer,GCP Data Engineer,Anywhere,via ZipRecruiter,Contractor and Temp work,True,Georgia,2023-11-07 14:01:59,False,False,United States,NaN,NaN,NaN,smart folks inc,"['python', 'sql', 'gcp']","{'cloud': ['gcp'], 'programming': ['python', '..."
6,Senior Data Engineer,Senior Data Engineer - GCP Cloud,"Dearborn, MI",via LinkedIn,Full-time,False,"Florida, United States",2023-03-27 13:18:18,False,False,United States,NaN,NaN,NaN,"Miracle Software Systems, Inc","['sql', 'python', 'java', 'sql server', 'gcp',...","{'cloud': ['gcp', 'bigquery'], 'databases': ['..."
9,Data Scientist,Data Scientist II,Anywhere,via ZipRecruiter,Full-time,True,"New York, United States",2023-04-23 13:02:57,False,False,United States,NaN,NaN,NaN,"Radwell International, LLC","['sql', 'python', 'r', 'mongodb', 'mongodb', '...","{'analyst_tools': ['excel'], 'cloud': ['azure'..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
785673,Data Scientist,Asset Management Data Scientist,"New York, NY",via Trabajo.org,Full-time,False,"New York, United States",2023-10-16 06:01:53,True,False,United States,NaN,NaN,NaN,JPMorgan Chase & Co.,NaN,NaN
785685,Data Analyst,Data Analyst,"Indianapolis, IN",via Trabajo.org,Full-time,False,"Illinois, United States",2023-07-16 06:01:27,True,False,United States,NaN,NaN,NaN,"Aara Technologies, Inc","['sql', 'excel']","{'analyst_tools': ['excel'], 'programming': ['..."
785692,Data Scientist,Data Scientist- Hybrid Work Location,"Dayton, OH",via Ai-Jobs.net,Full-time,False,"Illinois, United States",2023-04-21 06:02:31,False,True,United States,year,157500.0,NaN,Tenet3,"['python', 'matlab', 'r']","{'programming': ['python', 'matlab', 'r']}"


- Running this, and Printing it out, we can see that yes it did in fact Filter it for the United States but we have a Problem here
    - The `job_index` here doesn't go in Iterations of 1 by 1 or sequential Order
    - Now because we've extracted out these Records based on their Countries, we have missing Values in here...
- Luckily there's a Method here, that we can use that can help us out! Ofc there is!
    - We can use the `reset_index()` Method for this
        - `pandas.DataFrame.reset_index`
            - This Resets the Index
            - There's a host of different Parameteres we can use in this...
            - But we are just going to focus on one, the `inplace`
            - **Inplace**: This is whether to Modify the DataFrame rather than Creating a New one. By Default it's Set to *False* but we are going to Set it to *True* so we don't have to Create a New Variable
- So with our `df_usa` Data Frame we can invoke the `reset_index()` Method and then inside of it, we specify the `inplace=True`
    - Then we Inspect to see what happens!

In [16]:
# Running the reset_index Method on our Data Frame
df_usa.reset_index(inplace=True)

# Printing the Data Frame to Inspect it and show our work
df_usa

,job_index,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
0,0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
1,3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
2,5,Data Engineer,GCP Data Engineer,Anywhere,via ZipRecruiter,Contractor and Temp work,True,Georgia,2023-11-07 14:01:59,False,False,United States,NaN,NaN,NaN,smart folks inc,"['python', 'sql', 'gcp']","{'cloud': ['gcp'], 'programming': ['python', '..."
3,6,Senior Data Engineer,Senior Data Engineer - GCP Cloud,"Dearborn, MI",via LinkedIn,Full-time,False,"Florida, United States",2023-03-27 13:18:18,False,False,United States,NaN,NaN,NaN,"Miracle Software Systems, Inc","['sql', 'python', 'java', 'sql server', 'gcp',...","{'cloud': ['gcp', 'bigquery'], 'databases': ['..."
4,9,Data Scientist,Data Scientist II,Anywhere,via ZipRecruiter,Full-time,True,"New York, United States",2023-04-23 13:02:57,False,False,United States,NaN,NaN,NaN,"Radwell International, LLC","['sql', 'python', 'r', 'mongodb', 'mongodb', '...","{'analyst_tools': ['excel'], 'cloud': ['azure'..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
206287,785673,Data Scientist,Asset Management Data Scientist,"New York, NY",via Trabajo.org,Full-time,False,"New York, United States",2023-10-16 06:01:53,True,False,United States,NaN,NaN,NaN,JPMorgan Chase & Co.,NaN,NaN
206288,785685,Data Analyst,Data Analyst,"Indianapolis, IN",via Trabajo.org,Full-time,False,"Illinois, United States",2023-07-16 06:01:27,True,False,United States,NaN,NaN,NaN,"Aara Technologies, Inc","['sql', 'excel']","{'analyst_tools': ['excel'], 'programming': ['..."
206289,785692,Data Scientist,Data Scientist- Hybrid Work Location,"Dayton, OH",via Ai-Jobs.net,Full-time,False,"Illinois, United States",2023-04-21 06:02:31,False,True,United States,year,157500.0,NaN,Tenet3,"['python', 'matlab', 'r']","{'programming': ['python', 'matlab', 'r']}"
206290,785703,Data Analyst,Data Analyst - CRYPTOGRAPHY - Full-time,Anywhere,via Snagajob,Full-time,True,"Texas, United States",2023-01-28 06:02:55,True,False,United States,NaN,NaN,NaN,Axelon Services Corporation,NaN,NaN


- So now with this, we've kept our `job_index` in here but now we have this New Numerical Index, we can see it goes down to 206,292 Rows
- So it has what we want in this, the Index has been Reset!
    - Now technically, we can go through now and drop the `job_index` if we wanted but it can be argued that it should be kept
        - Because let's say in the future, when we learn about Merge Data Frames and we want to use an Index to Merge it back, we can then use that in this case
        - So we will keep it for the time being

- The next Method to learn about is regarding Setting the Index
    - For this, we can use the `set_index()` Method
        - `pandas.DataFrame.set_index`
            - This Sets the Data Frame Index unsing an existing Column
            - So we can take an existing Column and just move it over to the Index
- For the recently Altered Data Frame, let's say for some reason, we wanted to go back to that `job_index`
    - Really we can go to any Column, but we are going to go back to `job_index`
- Start by specifying a Data Frame, using the `set_index` Method
    - First thing to specify is: Keys - Specifically in this case, Series
        - We are going to Specify, that Column Title that we want to Set as the Index - So in our case, it's `job_index` here
        - Similar to before, we are going to set the `inplace=True`
- **Note**: We can either do it this way with the inplace, or set the Variable equal to itself. Both do the same thing but one is shorter than the other
    - Inplace just makes it so we Modify the Current Data Frame and we don't have to Create another one (setting it equal to itself) 

In [17]:
# Running the set_index Method on our Data Frame
df_usa.set_index('job_index', inplace=True)

# Printing the Data Frame to Inspect it and show our work
df_usa

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
job_index,,,,,,,,,,,,,,,,,
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
5,Data Engineer,GCP Data Engineer,Anywhere,via ZipRecruiter,Contractor and Temp work,True,Georgia,2023-11-07 14:01:59,False,False,United States,NaN,NaN,NaN,smart folks inc,"['python', 'sql', 'gcp']","{'cloud': ['gcp'], 'programming': ['python', '..."
6,Senior Data Engineer,Senior Data Engineer - GCP Cloud,"Dearborn, MI",via LinkedIn,Full-time,False,"Florida, United States",2023-03-27 13:18:18,False,False,United States,NaN,NaN,NaN,"Miracle Software Systems, Inc","['sql', 'python', 'java', 'sql server', 'gcp',...","{'cloud': ['gcp', 'bigquery'], 'databases': ['..."
9,Data Scientist,Data Scientist II,Anywhere,via ZipRecruiter,Full-time,True,"New York, United States",2023-04-23 13:02:57,False,False,United States,NaN,NaN,NaN,"Radwell International, LLC","['sql', 'python', 'r', 'mongodb', 'mongodb', '...","{'analyst_tools': ['excel'], 'cloud': ['azure'..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
785673,Data Scientist,Asset Management Data Scientist,"New York, NY",via Trabajo.org,Full-time,False,"New York, United States",2023-10-16 06:01:53,True,False,United States,NaN,NaN,NaN,JPMorgan Chase & Co.,NaN,NaN
785685,Data Analyst,Data Analyst,"Indianapolis, IN",via Trabajo.org,Full-time,False,"Illinois, United States",2023-07-16 06:01:27,True,False,United States,NaN,NaN,NaN,"Aara Technologies, Inc","['sql', 'excel']","{'analyst_tools': ['excel'], 'programming': ['..."
785692,Data Scientist,Data Scientist- Hybrid Work Location,"Dayton, OH",via Ai-Jobs.net,Full-time,False,"Illinois, United States",2023-04-21 06:02:31,False,True,United States,year,157500.0,NaN,Tenet3,"['python', 'matlab', 'r']","{'programming': ['python', 'matlab', 'r']}"


In [15]:
# The below does the exact same thing as the Cell above, I'm just adding it here to show that it can be done multiple ways and both accomplish the same thing
df_usa = df_usa.set_index('job_index')

df_usa

,job_title_short,job_title,job_location,job_via,job_schedule_type,job_work_from_home,search_location,job_posted_date,job_no_degree_mention,job_health_insurance,job_country,salary_rate,salary_year_avg,salary_hour_avg,company_name,job_skills,job_type_skills
job_index,,,,,,,,,,,,,,,,,
0,Senior Data Engineer,Senior Clinical Data Engineer / Principal Clin...,"Watertown, CT",via Work Nearby,Full-time,False,"Texas, United States",2023-06-16 13:44:15,False,False,United States,NaN,NaN,NaN,Boehringer Ingelheim,NaN,NaN
3,Data Engineer,LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...,"San Antonio, TX",via Diversity.com,Full-time,False,"Texas, United States",2023-07-04 13:01:41,True,False,United States,NaN,NaN,NaN,Southwest Research Institute,"['python', 'c++', 'java', 'matlab', 'aws', 'te...","{'cloud': ['aws'], 'libraries': ['tensorflow',..."
5,Data Engineer,GCP Data Engineer,Anywhere,via ZipRecruiter,Contractor and Temp work,True,Georgia,2023-11-07 14:01:59,False,False,United States,NaN,NaN,NaN,smart folks inc,"['python', 'sql', 'gcp']","{'cloud': ['gcp'], 'programming': ['python', '..."
6,Senior Data Engineer,Senior Data Engineer - GCP Cloud,"Dearborn, MI",via LinkedIn,Full-time,False,"Florida, United States",2023-03-27 13:18:18,False,False,United States,NaN,NaN,NaN,"Miracle Software Systems, Inc","['sql', 'python', 'java', 'sql server', 'gcp',...","{'cloud': ['gcp', 'bigquery'], 'databases': ['..."
9,Data Scientist,Data Scientist II,Anywhere,via ZipRecruiter,Full-time,True,"New York, United States",2023-04-23 13:02:57,False,False,United States,NaN,NaN,NaN,"Radwell International, LLC","['sql', 'python', 'r', 'mongodb', 'mongodb', '...","{'analyst_tools': ['excel'], 'cloud': ['azure'..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
785673,Data Scientist,Asset Management Data Scientist,"New York, NY",via Trabajo.org,Full-time,False,"New York, United States",2023-10-16 06:01:53,True,False,United States,NaN,NaN,NaN,JPMorgan Chase & Co.,NaN,NaN
785685,Data Analyst,Data Analyst,"Indianapolis, IN",via Trabajo.org,Full-time,False,"Illinois, United States",2023-07-16 06:01:27,True,False,United States,NaN,NaN,NaN,"Aara Technologies, Inc","['sql', 'excel']","{'analyst_tools': ['excel'], 'programming': ['..."
785692,Data Scientist,Data Scientist- Hybrid Work Location,"Dayton, OH",via Ai-Jobs.net,Full-time,False,"Illinois, United States",2023-04-21 06:02:31,False,True,United States,year,157500.0,NaN,Tenet3,"['python', 'matlab', 'r']","{'programming': ['python', 'matlab', 'r']}"


- We can see that we now have that `job_index` back in here and Inspecting it, we can see that it just Drops what we previously had as the Index Column (The unnamed Index Column)
    - It just Completely Drops it and now it's not there and ceases to exist
    - This is something to think about whenever you use this Method, is that we lose Data sometimes

- The last Method to cover in here is for Sorting Indexes
    - For this, we can use the `sort_index()` Method
        - `pandas.DataFrame.sort_index`
            - Returns a Data Frame Sorted by a Label 
            - In this case, we don't necessarily need to Specify the Index because, well it's Sorting the Index
- Now with the Operations we just did on our current Data Frame (df_usa) these Index Values remained in order so Sorting the Index here... there's not really a use case for it
- However there is a use case for this whenever we are Pivoting or Grouping Data
- Specifically, let's look at the previous example we did, where we Pivoted the `job_title_short` Column to find the Median Salary
    - And just for fun, let's add in the Min and the Max

In [39]:
# Pivoting the `job_title_short` Column to find Median, Min and Max
median_min_max_pivot = df_usa.pivot_table(index='job_title_short', values='salary_year_avg', aggfunc=['median', 'min', 'max'])

# Printing it out to show our work
median_min_max_pivot

,median,min,max
,salary_year_avg,salary_year_avg,salary_year_avg
job_title_short,,,
Business Analyst,90000.0,35000.0,387460.0
Cloud Engineer,116100.0,42000.0,221844.0
Data Analyst,90000.0,25000.0,375000.0
Data Engineer,125000.0,23496.0,525000.0
Data Scientist,130000.0,30000.0,960000.0
Machine Learning Engineer,150000.0,44408.5,315000.0
Senior Data Analyst,110000.0,35360.0,425000.0
Senior Data Engineer,150000.0,45000.0,375000.0


- Running this, we can see that in fact, the Index is `job_title_short` and it has all these different Values in it
- We could Sort by the Median Values and we saw this previously using the `.sort_values()` Method
    - For this:
        - We call `median_min_max_pivot`
        - Then use the `sort_values()` Method
        - Need to specify `by=`
            - This actually has 2 levels of Column Titles: **Median** and then `salary_year_avg`. So we will need to Provide this as a Tuple which is the Titles, in Parenthesis
            - Need to provide the title of 'median' separated by a comma and then the title 'salary_year_avg'
            - Then need to specify we want it in Ascending Order so we need to add `ascending=False` to the end

In [40]:
# Sorting the Pivot
median_min_max_pivot.sort_values(by=('median','salary_year_avg'), ascending=False)

,median,min,max
,salary_year_avg,salary_year_avg,salary_year_avg
job_title_short,,,
Senior Data Scientist,155000.0,55000.0,475000.0
Machine Learning Engineer,150000.0,44408.5,315000.0
Senior Data Engineer,150000.0,45000.0,375000.0
Data Scientist,130000.0,30000.0,960000.0
Software Engineer,130000.0,48982.0,375000.0
Data Engineer,125000.0,23496.0,525000.0
Cloud Engineer,116100.0,42000.0,221844.0
Senior Data Analyst,110000.0,35360.0,425000.0


- Now we have it in Ascending Order
- That was just to demonstrate how to do Median
- Well now similar to this, we can use the `.sort_index()` Method in order to Sort these Values
    - Specify our Variable: `median_min_max_pivot`
    - Then we do the `.sort_index()`

In [43]:
# Running the sort_index Method on our Data Frame
median_min_max_pivot.sort_index()

,median,min,max
,salary_year_avg,salary_year_avg,salary_year_avg
job_title_short,,,
Business Analyst,90000.0,35000.0,387460.0
Cloud Engineer,116100.0,42000.0,221844.0
Data Analyst,90000.0,25000.0,375000.0
Data Engineer,125000.0,23496.0,525000.0
Data Scientist,130000.0,30000.0,960000.0
Machine Learning Engineer,150000.0,44408.5,315000.0
Senior Data Analyst,110000.0,35360.0,425000.0
Senior Data Engineer,150000.0,45000.0,375000.0


- We now get these Values Sorted and they are Sorted here, in Alphabetical Order
- If we wanted to make sure that we Modified the current Data Frame that we have all we need to do is add `inplace=True`

In [45]:
# Running the sort_index Method on our Data Frame with inplace
median_min_max_pivot.sort_index(inplace=True)

# Printing it out
median_min_max_pivot

,median,min,max
,salary_year_avg,salary_year_avg,salary_year_avg
job_title_short,,,
Business Analyst,90000.0,35000.0,387460.0
Cloud Engineer,116100.0,42000.0,221844.0
Data Analyst,90000.0,25000.0,375000.0
Data Engineer,125000.0,23496.0,525000.0
Data Scientist,130000.0,30000.0,960000.0
Machine Learning Engineer,150000.0,44408.5,315000.0
Senior Data Analyst,110000.0,35360.0,425000.0
Senior Data Engineer,150000.0,45000.0,375000.0


- One thing to Note, you could technically, put inside the Sort the `job_title_short` (instead of the Tuple) since it does have a Name
    - EG: `median_min_max_pivot.sort_values(by='job_title_short', ascending=False)`
- And whenever you Run it, it would Run it by the Name, in Descending Order but it can still work
- So then the question is, when would this `.sort_index()` be a use case then?
    - And that's when, maybe the Index doesn't have a Column Title
    - Which is what we saw at the Start of this Lesson, before we Assigned a Column Title to the Index, that's when this would be Applicable
    - When we first Ran the Data Frame and we didn't have a Name for it, we definitely want to use the `.sort_index()` Method in this case when it doesn't have a name
- So being able to Manipulate Indexes is going to get you out of trouble from time to time and also keep your Data in a very logical and also Structured Manner